Change GADT constrainer to consider all parents of the scrutinee type when upcasting it

Linyxus opened the pull request.

abgruszecki commented on compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
-          scrut.firstParent
+          // consider all parents
+          val parents = scrut.parents
+          val andType = trace(i"andType of scrut", gadts) {
```
This trace doesn't seem overall very useful, could you remove it?

abgruszecki commented on compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
+          val andType = trace(i"andType of scrut", gadts) {
+            buildAndType(parents)
+          }
+          constrainPatternType(pat, andType)
```
Can you check if `andType` exists, like in `case _` below?

abgruszecki commented on compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
-        case _ => NoType
+          val allSyms = allParentsSharedWithPat(tycon, tycon.symbol.asClass)
+          val baseClasses = allSyms map scrut.baseType
+          val andType = trace(i"andType of scrut", gadts) {
```
Ditto about the trace.

abgruszecki commented on compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
+          val andType = trace(i"andType of scrut", gadts) {
+            buildAndType(baseClasses)
+          }
+          constrainPatternType(pat, andType)
```
Ditto about `andType` existing.

abgruszecki reviewed (commented):
Good news, we can merge this PR in time for 3.0.0. Some tweaks would be nice, and then it's good to go!

Linyxus commented:
Thank you for your review! I have removed unnecessary traces and added the existence checking. It's really good that we can include this feature in 3.0.0. :)
